### DISCUSSION TITLE
How to generate clients for different services

### DISCUSSION THREAD

**Discussion by tobiaskahl** [Netflix/dgs-framework#746] [General]
I need clients for multiple services.

duplicating the "generatedJava"-Task does not work, because only each task run is deleting the older ones.
Including multiple schema files will cause duplicate variable names in DgsConstants.kt.

`DgsConstants.kt: (112, 17): Conflicting declarations: public object QUERY, public object QUERY`

`generateJava {`
`        schemaPaths = mutableListOf(`
`            "${rootDir}/academy-service/src/main/resources/graphql/schema.graphqls",`
`            "${rootDir}/user-service/src/main/resources/graphql/schema.graphqls",`
`        )`
`        packageName = "de.example.service.client.graphql"`
`        generateClient = true`
`    }`

thanks for some tips

**Comment by srinivasankavitha** [comment]
Are you able to separate your services into different modules, each with it's own build.gradle file? Then you can define a separate task for each service, which will solve your issue. That is how we address this issue for multiple services in the same project.

**Reply by tobiaskahl** [reply]
Hi ,
would be possible, but would unnecessarily increase the number of modules.
But thanks for the suggestion of the "workaround", which i may use if i can not fix this.

**Reply by srinivasankavitha** [reply]
Unless you modify your schema such that codegen can treat all the schema files as part of one schema, i.e. using `extend type Query` etc., you will have this issue. I'm not sure if that would be correct, but depends on your use case.

**Reply by tobiaskahl** [reply]
I will try this dirty hack:

`doLast { Files.delete(Paths.get("$buildDir/generated/${packageName.replace(".", File.separator)}/DgsConstants.kt")) }`

i doesn't know if this file is needed internally. I doesn't need ist for my clients.

**Reply by paulbakker** [reply]
It's not needed internally. 
Generally the recommendation is to put each "client" into a separate module though, because mixing the schemas for multiple clients and/or server doesn't really make sense.

**Reply by HeenaChawla21** [reply]
I tried creating different tasks in the following way and also deleted the DGSConstants file but it still provides this error :
    There are problems with the GraphQL Schema:
         * 'Query' type [@5:1] tried to redefine existing 'Query' type [@1:1]
         * 'BigDecimal' type [@81:1] tried to redefine existing 'BigDecimal' type [@59:1]

`generateJava {
    generateClient = true
    schemaPaths = ["****"]
    packageName = "****folder1"
    typeMapping = ["BigDecimal":"java.math.BigDecimal", "Long":"java.lang.Long","DateTime":"java.time.OffsetDateTime"]
    includeQueries = ["****"]

    doLast {
        delete files("${project.buildDir.absolutePath}/generated/sources/dgs-codegen/*****/folder1/DgsConstants.java")
    }
}
tasks.register("generateCodegen",com.netflix.graphql.dgs.codegen.gradle.GenerateJavaTask) {
    packageName = "****folder2"
    schemaPaths = ["*****"]
    includeQueries = ["*****"]
    typeMapping = ["BigDecimal":"java.math.BigDecimal", "Long":"java.lang.Long","DateTime":"java.time.OffsetDateTime"]
    generateClient = true

    doLast {
        delete files("${project.buildDir.absolutePath}/generated/sources/*****/folder2/DgsConstants.java")
    }
}`

**Comment by GrantGochnauer** [comment]
Hello - 

We are running into the same problem.

Consider this scenario:
* We have 2 independently deployed services using DGS with GraphQL schemas
* We have a single application that needs to make calls to both DGS services that have been deployed. Each have their own schema
* In my application, I want to generate all the DGS client code so that I can make calls into each respective service.

I too am running into errors like:
```
***************************
APPLICATION FAILED TO START
***************************

Description:

There are problems with the GraphQL Schema:
	 * 'Query' type [@1:1] tried to redefine existing 'Query' type [@1:1]
	 * 'Mutation' type [@67:1] tried to redefine existing 'Mutation' type [@72:1]
	 * 'SortBy' type [@209:1] tried to redefine existing 'SortBy' type [@808:1]
	 * 'Audit' type [@218:1] tried to redefine existing 'Audit' type [@734:1]
	 * 'SortableField' type [@228:1] tried to redefine existing 'SortableField' type [@744:1]
	 * 'OrderDirection' type [@237:1] tried to redefine existing 'OrderDirection' type [@749:1]
	 * 'UUID' type [@242:1] tried to redefine existing 'UUID' type [@817:1]
	 * 'JSON' type [@243:1] tried to redefine existing 'JSON' type [@818:1]
	 * 'DateTime' type [@244:1] tried to redefine existing 'DateTime' type [@819:1]
	 * 'Url' type [@245:1] tried to redefine existing 'Url' type [@820:1]
	 * 'Size' type [@247:1] tried to redefine existing directive 'Size' type [@822:1]
```

We are using Gradle and I have broken out the schema generation into their own gradle sub projects whose only responsibility is to generate the schema. For example:
```
plugins {
    id 'com.netflix.dgs.codegen' version '6.1.4'
}


tasks.named('generateJava') {
    schemaPaths = ["${projectDir}/src/main/resources/schema/iam"]
    packageName = 'com.xxx.cloud.iam.api' 

    typeMapping = ["Url"               : "java.net.URL",
                   "DateTime"          : "java.time.OffsetDateTime",
                   "JSON"              : "java.util.Map",
                   "Locale"            : "java.util.Locale",
                   "Long"              : "java.lang.Long",
                   "DocumentConnection": "graphql.relay.SimpleListConnection<Document>",
                   "UUID"              : "java.util.UUID"]
}
```
Then in my main gradle project, I have this:
`implementation project(':iam-api-schema')` that pulls in my sub project.

However the error is thrown from `DgsAutoConfiguration` class  because it can't instantiate the schema due to the overlapping schema definitions. I should note that each of the sub projects, with their own schemas, use different java packages. I was hoping that by having a different package, DGS would be able to handle schemas at the "package" level instead of being global.

Thoughts?

**Reply by GrantGochnauer** [reply]
@paulbakker Any thoughts on above - I'm trying to follow the recommendation of using different modules per your suggestion. Thanks!

**Reply by GrantGochnauer** [reply]
It also seems like DGS is anticipating use of multiple schemas as written in `DgsSchemaProvider.kt` there is code on line 161 that says:
```
        var mergedRegistry = if (schema == null) {
            val hasDynamicTypeRegistry = dgsComponents.any { it.annotatedMethods<DgsTypeDefinitionRegistry>().any() }
            val readerBuilder = MultiSourceReader.newMultiSourceReader()
                .trackData(false)
            for (schemaFile in findSchemaFiles(hasDynamicTypeRegistry)) {
                readerBuilder.reader(schemaFile.inputStream.reader(), schemaFile.filename)
                // Add a reader that inserts a newline between schema files to avoid issues when
                // the source files aren't newline-terminated.
                readerBuilder.reader("\n".reader(), "newline")
            }
            SchemaParser().parse(readerBuilder.build())
        } else {
            SchemaParser().parse(schema)
        }

```

**Comment by manjitsinghb** [comment]
Here is how we generate clients for 2 sub graphs in the same gradle build file
we had 2 sub graphs and we had conflicting schema names.

```
tasks.register('subgraph1', com.netflix.graphql.dgs.codegen.gradle.GenerateJavaTask) {
    schemaPaths = ["${projectDir}/src/main/resources/graph/schema/graph1.graphqls"]
    packageName = 'com.project.pojo.graphql1' // The package name to use to generate sources
    typeMapping = [
            "_Any": "java.util.Optional"
    ]
    generateClient = true // Enable generating the type safe query API
}

tasks.register('subgraph2', com.netflix.graphql.dgs.codegen.gradle.GenerateJavaTask) {
    schemaPaths = ["${projectDir}/src/main/resources/graph/schema/graph2.graphqls"]
    packageName = 'com.project.pojo.graphql2 // The package name to use to generate sources
    typeMapping = [
            "_Any": "java.util.Optional"
    ]
    generateClient = true // Enable generating the type safe query API
}

tasks.named("compileJava") {
    dependsOn(tasks.named("subgraph1"))
    dependsOn(tasks.named("subgraph2"))
}

```
Works as the pojos are generated in different  java packages and  conflict free
